### ISSUE TITLE
Improve dropdown button focus state

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/almanac.httparchive.org#892]
When the language switcher is focused, the border and text turn yellow, and there is also a blue/white outline:

![image](https://user-images.githubusercontent.com/1120896/85237809-85204500-b3f7-11ea-84a7-bfb337d189d4.png)

As long as the element changes appearance to yellow when focused, I don't think we need the outline.

One other minor style improvement might be to also change the dropdown arrow to yellow when focused. Playing with this in devtools, I'm not sure how easy this is given that the arrow is styled as an `::after` pseudo-element on the dropdown's parent element. Open to ideas.

cc @HTTPArchive/developers

**Comment by bazzadp** [comment]
> When the language switcher is focused, the border and text turn yellow, and there is also a blue/white outline:
> 
> ![image](https://user-images.githubusercontent.com/1120896/85237809-85204500-b3f7-11ea-84a7-bfb337d189d4.png)
> 
> As long as the element changes appearance to yellow when focused, I don't think we need the outline.

Hmmm... this is a Chrome specific thing (I believe they changed it recently). If you look at Firefox and Safari, they don't do this (though neither has been good with focus states IMHO!). I weary to change focus states as presume Chrome knows better than us, and this recent change has gone down well in accessibility community.

It's also a bigger issue than just the select:

![Focus state on Methodology link](https://user-images.githubusercontent.com/10931297/85253355-e37c1080-b455-11ea-9fe9-a2aa66ec3dfe.png)

I presume the square, rather than rounded corners is what's making it especially jarring for you on the select? Unfortunately there's [no way of changing that at present](https://bugs.chromium.org/p/chromium/issues/detail?id=81556&q=outline%20border-radius&can=2) without removing or re-implementing the whole focus state manually - which would look odd in non-Chrome browsers with different focus styling.

> One other minor style improvement might be to also change the dropdown arrow to yellow when focused. Playing with this in devtools, I'm not sure how easy this is given that the arrow is styled as an `::after` pseudo-element on the dropdown's parent element. Open to ideas.

Very easy, by changing this:

```
.language-switcher:hover,
.year-switcher:hover {
  color: #F7F779;
}
```

to this:

```
.language-switcher:hover,
.year-switcher:hover,
.language-switcher::after:hover,
.year-switcher::after:hover {
  color: #F7F779;
}
```

Will include in pull request for #815 if not one else has tackled before then.

**Comment by catalinred** [comment]
Indeed, the focus outline styles changed with the latest Chrome. Usually, as you already know, we can remove the default focus styles only when we provide an alternative. 
AFAIK, right now we're not using a different `:focus` style, so I wouldn't remove the defaults yet.

---

The native `select` is wrapped within another elem `.language-switcher` in order to be able to use that pseudo-element as a dropdown arrow. I did use `:focus-within` as a progressive enhancement to replicate the focus style on the native `select` element.

Besides toggling `:focus` explicitly on the `select` in DevTools' "Force element state", I wasn't able to replicate the "white dropdown arrow" when focusing on this custom select, neither with the Tab key nor with the mouse. Can you give more details on how you encountered this behavior?

**Comment by bazzadp** [comment]
Yeah I can't repeat it now either on production. But am sure I could earlier.... Odd. Either that or my eyes were playing tricks on me!

**Comment by rviscomi** [comment]
I think the fix for this is:

```css
.language-switcher:focus-within,
.year-switcher:focus-within,
.language-switcher select:focus,
.year-switcher select:focus {
  border-color: #f7f779;
  color: #f7f779;
  outline: none;
}
```

Where the outline is disabled entirely. I think the border color change is sufficient.

**Comment by bazzadp** [comment]
I think that's a fine change for those using the mouse, but less than ideal for keyboard users as when you tab into it as its very subtle change and, give our colour-scheme, it doesn't really differentiate the focused item unless you happen to notice it change.

And when you compare keyboard focus for the logo:

![Logo with focus](https://user-images.githubusercontent.com/10931297/87254478-ef6a5980-c47a-11ea-8e3d-1fb64854b72b.png)

Other text:

![Header link with focus](https://user-images.githubusercontent.com/10931297/87254487-fbeeb200-c47a-11ea-81bc-57fd186f9f30.png)

And the buttons:

![Button link with focus](https://user-images.githubusercontent.com/10931297/87254498-0e68eb80-c47b-11ea-8203-bf31b5f6862a.png)

All follow the same convention so with this change, the selects now look out of place with your suggested change.

You could make the focus style more obvious (e.g. blue colour), but with different browsers doing different things, it's going to be a tough call (impossible?) to not make it look consistent.

I'm guessing your main issue with this is that you see it on mouse click as well, whereas the other states you don't see this focus state on the other items as once you click on something, you navigate away? Personally it doesn't bother me - my focus (pardon the pun!) is on the drop down contents at that point.

Only doing your solution on `:hover` *mostly* solves your problem for mouse without ruining the keyboard experience:

```css
.language-switcher:focus-within,
.year-switcher:focus-within,
.language-switcher select:focus,
.year-switcher select:focus {
  border-color: #f7f779;
  color: #f7f779;
}

.language-switcher select:hover,
.year-switcher select:hover {
  outline: none;
}
```

Now you only see the ugly blue thing if you open the menu, then click away without selecting an item. Not ideal but should address the majority of use cases.

However I presume it only works on desktop and this is also an issue on mobile Chrome? Don't have an Android phone to test but guessing would still be an issue there?

Which brings me to my next point - Safari does this properly:

![Safari focus ring respects border-radius](https://user-images.githubusercontent.com/10931297/87255350-e630bb00-c481-11ea-9039-51711ff06e40.png)

And the reason that Chrome doesn't is due to this very old bug: https://bugs.chromium.org/p/chromium/issues/detail?id=81556. Guessing that won't be solved any time soon given it was opening 9 years ago, unless the recent focus change upsets more people like it seems to with you and that forces Chrome into action.

So, in summary, I don't think we should change this to be honest. But I certainly won't stop you submitting your fix if it really is upsetting you - been there and can understand! I just think it's better for accessibility to leave as is, and not seeing the big deal with it at present to be honest.

**Comment by bazzadp** [comment]
By complete coincidence I just came across `:focus-visible` tonight, which is another way to solve this: https://css-tricks.com/almanac/selectors/f/focus-visible/

Unfortunately not available in Chrome yet, but there has been recent activity on it: https://bugs.chromium.org/p/chromium/issues/detail?id=817199

**Comment by bazzadp** [comment]
I'd rather explore other options. The focus ring isn't just about changing the color when you move, but also clearer differentiating it from other items. Color alone doesn't do thst for us as, without the focus-ring, the focus-ed year select looks the same as the Star Exploring button.

What do you think about just making it a yellow focus ring?

```css
:focus {
  outline: #f7f779 dotted 1px;
  outline-offset: 2px;
}
```

![Screenshot of various components with and without focus outline styling](https://user-images.githubusercontent.com/10931297/87290222-a0153f00-c4f5-11ea-8738-be9c644e8a90.png)

**Comment by bazzadp** [comment]
Released some versions to allow us to test:

- [Current production](https://almanac.httparchive.org/en/2019/) - blue ring on focus
- [Test Version 1](https://20200714t080417-dot-webalmanac.uk.r.appspot.com) - removes focus ring entirely from header, footer and buttons
- [Test Version 2](https://20200714t080741-dot-webalmanac.uk.r.appspot.com) - leaves focus ring but makes it dotted yellow

**Comment by ericwbailey** [comment]
>As long as the element changes appearance to yellow when focused, I don't think we need the outline.

I would fail this under [WCAG SC 1.4.1 Use of Color](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-without-color.html), as color alone is not enough to indicate state change. A good alternative to this would be to more obviously adjust the visual presentation. Inverting the hollow treatment to be filled might be a good approach.

I would also advise not removing default focus indicators unless the focus state is heavily apparent. Thin dotted outlines are oftentimes not sufficient for low vision users.

**Comment by bazzadp** [comment]
Yeah that was my concern. Though do get the concern around the blue focus ring in Chrome not being a great fit for our brand colours, so I was OK with changing it to yellow. Could go with either solid and/or increase the width of the outline to address any concerns with that if being too easily missed.

However, I've also spotted it's very faint on white backgrounds:

![Focus ring difficult to see against white background](https://user-images.githubusercontent.com/10931297/87445179-a12a9700-c5ef-11ea-8086-9b2a036856cc.png)

this is why sticking with tried and tested browser defaults are a good thing! 😀

> Inverting the hollow treatment to be filled might be a good approach.

Can you explain what you mean by this? We invert the colours on the "Start Exploring" button on the home page on focus for example, but not on the select drop downs. However when tabbing to the controls if they are off-screen, you miss this colour change. Try it by tabbing through the home page now.

**Comment by ericwbailey** [comment]
Sure thing. I think this treatment would be effective:

<img width="338" alt="A button labeled 'Start exploring' with a focus state where the background has gone from transparent to filled." src="https://user-images.githubusercontent.com/634191/87446797-c9060480-c5c7-11ea-818c-44ad6653d3cb.png">

I think the treatment would be applicable for select drop-downs as well, in that you're visually indicating that the control has received focus and the down-facing chevron indicates the presence of dropdown functionality. There'd also be a nice little nod for cognitive accessibility concerns, in that the focus state would be more internally consistent for interactive elements.

**Comment by bazzadp** [comment]
Sorry still confused.

So on the site at the moment we have:

Unfocused:

![Unfocused is yellow](https://user-images.githubusercontent.com/10931297/87447469-8c9bce00-c5f2-11ea-9014-e463a72f7b38.png)


Hover:

![Hover is grey](https://user-images.githubusercontent.com/10931297/87447538-9b828080-c5f2-11ea-83cc-9b60a7ffb31f.png)


Tab-focused:
![Tab-focused is grey (same as Hover) plus blue hover ring](https://user-images.githubusercontent.com/10931297/87447576-a6d5ac00-c5f2-11ea-9cfc-c7bb2b3da4d4.png)

So we already have what you are suggesting? Though admittedly don't have this as strongly in the Year and Language selectors where we only change the border colour so should definitely do something there if removing the focus ring.

Anyway, that is why @rviscomi is advocating dropping the blue ring - as we already have a visual indicator (and strong one for button, though weaker one for selects). Though I'm less convinced because, as I say, if you start in header, then tab to this button it's off screen initially so you don't see the change in colour. So think we need a focus ring as well.

**Comment by bazzadp** [comment]
Also we alternate the colours, which is potentially a little confusing. So next button is this:

Unfocused:

![Dark blue filled button](https://user-images.githubusercontent.com/10931297/87448579-efda3000-c5f3-11ea-88a0-b5e8ce756a7d.png)

Hover:

![White filled button](https://user-images.githubusercontent.com/10931297/87448611-f9639800-c5f3-11ea-89ad-4972d5b2b113.png)

Tab focused:

![White filled button (same as hover) with focus ring](https://user-images.githubusercontent.com/10931297/87448660-05e7f080-c5f4-11ea-9caa-32f61b5847e6.png)

**Comment by obto** [comment]
@bazzadp pardon the D&D terminology, but this kind of gets into a RAW (rules as written) vs RAI (Rules as intended) debate, and I'm personally a Rules as Intended person.

The reason the focus ring is important, is it makes it much easier to navigate a site using ONLY the keyboard... which is the experience most a11y devices end up emulating. So as long as when you tab through the site and can clearly see where you are each step of the way, that's what matters.

**Comment by bazzadp** [comment]
Yes I totally agree that is what we are all aiming for. However, we're in some dispute as to whether the removing the focus ring still allows you to "clearly see where you are each step of the way" so asking for others expert opinion here. My vote is to keep it.

**Comment by ericwbailey** [comment]
+1 to keeping it.

**Comment by obto** [comment]
+1. It may not look the prettiest, but it does help make it a bit more obvious where you are.

**Comment by bazzadp** [comment]
This was fixed with #1214 by using `focus-visible` - now only get focus ring on keyboard navigation and not on click.

Currently available in Chrome Canary and coming to a Chrome stable in October. Firefox has it with prefix (that we don't support) and [Igalia are crowdfunding for Safari](https://www.igalia.com/open-prioritization/).
